src/NITTA/Synthesis/Method.hs

```diff
@@ -46,7 +46,6 @@ the endless synthesis process.
 stepLimit = 750 :: Int
 
 -- |The most complex synthesis method, which embedded all another. That all.
-stateOfTheArtSynthesisIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
```

Comment:
I made these changes because of a compilation error. There's `exploreSubForestVar` method in `Explore.hs` that had signature:
```
exploreSubForestVar ::
    (UnitTag tag, VarValTime v x t) =>
    Tree (TargetSystem (BusNetwork tag v x t) tag v x t) tag v x t ->
    STM [Tree (TargetSystem (BusNetwork tag v x t) tag v x t) tag v x t]
```
But after adding `++ map (decisionAndContext parent) (sBindPUOptions sState)` signature has change to:
```
exploreSubForestVar ::
    ( UnitTag tag
    , VarValTime v x t
    , ToJSON p
    , Viewable d DecisionView
    , SynthesisDecisionCls
        ( SynthesisState
            (TargetSystem (BusNetwork tag v x t) tag v x t)
            tag
            v
            x
            t
        )
        (TargetSystem (BusNetwork tag v x t) tag v x t)
        (BindPU tag) -- Why BindPU is here????
        d
        p
    , Typeable p
    , Show d
    ) =>
    Tree (TargetSystem (BusNetwork tag v x t) tag v x t) tag v x t ->
    STM [Tree (TargetSystem (BusNetwork tag v x t) tag v x t) tag v x t]
```

I spent a lot of time on this, so then I just went down the easier path.

Reply:
I think you can a little but change `SynthesisMethod` type. It will be better solution for haddock

Reply:
```diff
diff --git a/nitta.cabal b/nitta.cabal
index 7c95d803..c1b09356 100644
--- a/nitta.cabal
+++ b/nitta.cabal
@@ -60,6 +60,7 @@ library
       NITTA.Model.Problems.Refactor.OptimizeAccum
       NITTA.Model.Problems.Refactor.ResolveDeadlock
       NITTA.Model.Problems.ViewHelper
+      NITTA.Model.ProcessIntegrity
       NITTA.Model.ProcessorUnits
       NITTA.Model.ProcessorUnits.Accum
       NITTA.Model.ProcessorUnits.Broken
diff --git a/src/NITTA/Synthesis/Method.hs b/src/NITTA/Synthesis/Method.hs
index 839a3142..5cf8fe0e 100644
--- a/src/NITTA/Synthesis/Method.hs
+++ b/src/NITTA/Synthesis/Method.hs
@@ -31,7 +31,9 @@ import qualified Data.List as L
 import Data.Typeable
 import Debug.Trace
 import NITTA.Model.ProcessorUnits
+
 import NITTA.Model.TargetSystem
+import NITTA.Synthesis.Allocation ()
 import NITTA.Synthesis.Bind
 import NITTA.Synthesis.Explore
 import NITTA.Synthesis.Refactor
@@ -46,6 +48,7 @@ the endless synthesis process.
 stepLimit = 750 :: Int
 
 -- |The most complex synthesis method, which embedded all another. That all.
+stateOfTheArtSynthesisIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 stateOfTheArtSynthesisIO tree = do
     infoM "NITTA.Synthesis" $ "stateOfTheArtSynthesisIO: " <> show (sID tree)
     l1 <- simpleSynthesisIO tree
@@ -55,16 +58,19 @@ stateOfTheArtSynthesisIO tree = do
     return $ bestLeaf tree [l1, l2, l3, l4]
 
 -- |Schedule process by simple synthesis.
+simpleSynthesisIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 simpleSynthesisIO root = do
     infoM "NITTA.Synthesis" $ "simpleSynthesisIO: " <> show (sID root)
     lastObliviousNode <- obviousBindThreadIO root
     allBestThreadIO 1 lastObliviousNode
 
+smartBindSynthesisIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 smartBindSynthesisIO tree = do
     infoM "NITTA.Synthesis" $ "smartBindSynthesisIO: " <> show (sID tree)
     tree' <- smartBindThreadIO tree
     allBestThreadIO 1 tree'
 
+bestThreadIO :: (VarValTime v x t, UnitTag tag) => Int -> SynthesisMethod tag v x t
 bestThreadIO 0 node = return $ trace "bestThreadIO reach step limit!" node
 bestThreadIO limit tree = do
     subForest <- positiveSubForestIO tree
@@ -72,12 +78,14 @@ bestThreadIO limit tree = do
         [] -> return tree
         _ -> bestThreadIO (limit -1) $ maximumOn (score . sDecision) subForest
 
+bestStepIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 bestStepIO tree = do
     subForest <- positiveSubForestIO tree
     case subForest of
         [] -> error "all step is over"
         _ -> return $ maximumOn (score . sDecision) subForest
 
+obviousBindThreadIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 obviousBindThreadIO tree = do
     subForest <- positiveSubForestIO tree
     maybe (return tree) obviousBindThreadIO $
@@ -92,6 +100,7 @@ obviousBindThreadIO tree = do
             )
             subForest
 
+allBindsAndRefsIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 allBindsAndRefsIO tree = do
     subForest <-
         filter ((\d -> isBind d || isRefactor d) . sDecision)
@@ -105,6 +114,7 @@ refactorThreadIO tree = do
     maybe (return tree) refactorThreadIO $
         L.find (isRefactor . sDecision) subForest
 
+smartBindThreadIO :: (VarValTime v x t, UnitTag tag) => SynthesisMethod tag v x t
 smartBindThreadIO tree = do
     subForest <-
         filter ((\d -> isBind d || isRefactor d) . sDecision)
@@ -113,6 +123,7 @@ smartBindThreadIO tree = do
         [] -> return tree
         _ -> smartBindThreadIO $ maximumOn (score . sDecision) subForest
 
+allBestThreadIO :: (VarValTime v x t, UnitTag tag) => Int -> SynthesisMethod tag v x t
 allBestThreadIO (0 :: Int) tree = bestThreadIO stepLimit tree
 allBestThreadIO n tree = do
     subForest <- positiveSubForestIO tree
```

Do you know why it helps?